improver/utilities/extract_from_table.py

```diff
@@ -0,0 +1,175 @@
+    @staticmethod
+    def nearest_lower_index(
+        values: np.array, sorted_table_labels: np.array
+    ) -> np.array:
+        """Returns the index of the nearest lower label for every element of values.
+            Args:
+                values:
+                    Array of values to extract from table
+                table_label:
+                    An array of the labels along an axis of the table.
+            Returns:
+                An array of indices of the nearest lower label for each value in values.
+        """
+
+        sorted_index = np.searchsorted(sorted_table_labels, values, side="right") - 1
+
+        # If floating points are being used in the table_labels and values numpy may not
+        # recognise that the values are equal. This can lead to the wrong index being selected so
+        # we need to check if the values are within a tolerance of the table_labels and correct the
+        # index if necessary.
+        tol = 1e-5
+        condition = np.abs(np.array(sorted_table_labels)[:, None] - values) < tol
+        true_index, index = np.where(condition)
+
+        sorted_index[index] = true_index
+
+        # If value is below lowest table_label then index will be -1. We need to correct this to 0.
+        sorted_index = np.clip(sorted_index, 0, None)
+
+        return sorted_index
+
+    def extract_table_values(
+        self, table: DataFrame, columns_cube: iris.cube.Cube, row_cube: iris.cube.Cube
+    ) -> np.array:
+        """Extract values from the table based on the provided row and column cubes.
+            Args:
+                table:
+                    DataFrame representing the table from which values are extracted.
+                columns_cube:
+                    Cube used to index the columns of the table.
+                row_cube:
+                    Cube used to index the rows of the table.
+            Returns:
+                Array of values extracted from the table.
+        """
+        shape = columns_cube.shape
+
+        columns_data = columns_cube.data.flatten()
+        row_data = row_cube.data.flatten()
+        column_index = self.nearest_lower_index(
+            values=columns_data, sorted_table_labels=table.columns
+        )
+        row_index = self.nearest_lower_index(
+            values=row_data, sorted_table_labels=table.index
+        )
+
+        result = table.to_numpy()[row_index, column_index]
+        result = result.reshape(shape)
+        return result
+
+    def convert_dict_to_dataframe(self, table: dict) -> DataFrame:
+        """Converts a dictionary to a pandas DataFrame"""
+
+        table = DataFrame.from_dict(table)
+        table.columns = table.columns.astype(float)
+        table.index = table.index.astype(float)
+
+        table = table.reindex(sorted(table.columns), axis=1)
+        table = table.reindex(sorted(table.index), axis=0)
+
+        return table
+
+    def process(self, *cubes: Union[iris.cube.CubeList, iris.cube.Cube], table: dict):
```

Comment:
I think the method only accepts a list-like object with two cubes in it - anything else will raise an error.
```suggestion
    def process(self, *cubes: List[Cube, Cube], table: dict):
```

Reply:
Done